cpp/src/reductions/scan/scan_inclusive.cu

```diff
@@ -155,8 +157,207 @@ struct scan_dispatcher {
+          return row_index + 1;
+        }
+      }
+
+      return thrust::all_of(
+               thrust::seq,
+               thrust::make_counting_iterator<size_type>(0),
+               thrust::make_counting_iterator<size_type>(0) + d_order_by.num_child_columns(),
+               [row_index, d_order_by] __device__(size_type child_index) {
+                 column_device_view col = d_order_by.child(child_index);
+                 element_equality_comparator<nested_nulls> element_comparator{col, col, true};
+                 return cudf::type_dispatcher(
+                   col.type(), element_comparator, row_index, row_index - 1);
+               })
+               ? 0
+               : row_index + 1;
+    });
+}
+
+template <bool has_nulls>
+void generate_rank_comparisons(column_view const& order_by,
+                               mutable_column_view out,
+                               rmm::cuda_stream_view stream)
+{
+  auto d_order_by = column_device_view::create(order_by, stream);
+  element_equality_comparator<has_nulls> element_comparator(*d_order_by, *d_order_by, true);
+  thrust::tabulate(
+    rmm::exec_policy(stream),
+    out.begin<size_type>(),
+    out.end<size_type>(),
+    [type = d_order_by->type(), element_comparator] __device__(size_type row_index) {
+      return (row_index == 0 ||
+              !cudf::type_dispatcher(type, element_comparator, row_index, row_index - 1))
+               ? row_index + 1
+               : 0;
+    });
+}
+
+template <bool nested_nulls>
+void generate_dense_rank_struct_comparisons(column_view const& order_by,
+                                            mutable_column_view out,
+                                            rmm::cuda_stream_view stream)
+{
+  thrust::tabulate(
+    rmm::exec_policy(stream),
+    out.begin<size_type>(),
+    out.end<size_type>(),
+    [d_order_by = *column_device_view::create(order_by, stream),
+     has_nulls  = order_by.has_nulls()] __device__(size_type row_index) {
+      if (row_index == 0) {
+        return 1;
+      } else if (has_nulls) {
+        bool const lhs_is_null{d_order_by.is_null_nocheck(row_index)};
+        bool const rhs_is_null{d_order_by.is_null_nocheck(row_index - 1)};
+        if (lhs_is_null and rhs_is_null) {
+          return 0;
+        } else if (lhs_is_null != rhs_is_null) {
+          return 1;
+        }
+      }
+
+      return thrust::all_of(
+               thrust::seq,
+               thrust::make_counting_iterator<size_type>(0),
+               thrust::make_counting_iterator<size_type>(0) + d_order_by.num_child_columns(),
+               [row_index, d_order_by] __device__(size_type child_index) {
+                 column_device_view col = d_order_by.child(child_index);
+                 element_equality_comparator<nested_nulls> element_comparator{col, col, true};
+                 return cudf::type_dispatcher(
+                   col.type(), element_comparator, row_index, row_index - 1);
```

Comment:
One thing I want to point out is that, we don't need type_dispatcher for comarator. Just call the function, which will invoke type_dispatcher by itself.

```suggestion
                 return element_comparator(row_index, row_index - 1);
```

Reply:
This applies in many places in this PR.

Reply:
This is necessary because the `element_comparator` gives full output for anything that doesn't work with a `==` equality comparator. As a result, the compiler will give a `error: no instance of overloaded function "cudf::element_equality_comparator<has_nulls>::operator() [with has_nulls=true]" matches the argument list`